### PR TITLE
Incremental backup: do not error on empty backup

### PR DESCRIPTION
**Pull request by shlomi-noach**
## Description

An incremental backup backs up the delta of binary logs from last known good backup up to the point where the incremental backup is requested. An incremental backup flushes the binary logs and sees what binary log content needs to be backed-up.

It is possible for an incremental backup to be _empty_. This can happen when the database is sitting idly doing absolutely nothing from the last good backup to the point of incremental backup.

Up till now, this resulted with an error: `vterrors.Errorf(vtrpc.Code_FAILED_PRECONDITION, "no binary logs to backup (increment is empty)")`.

But this is an undesired behavior. It is possible for databases to sit idly. If someone were to set up a cronjob running an incremental backup at, say, every `1min` or every `5min`, then they'd have to deal with a lot of errors that really are not indicative of anything wrong per se in the database or in the backup process.

The decision to error on an empty backup was never formalized and was left as an arbitrary behavior. We now fomalize the behavior for an empty backup:

- An empty backup is possible for incremental backups only. It is irrelevant to full backups.
- An empty backup happens when there's no new GTIDs since last good backup, ie no content in the binary logs that needs to be copied.
- An empty backup exits with return code `0`, ie **not an error**.
- An empty backup clearly indicates `"backup is empty"` in its standard output (wording might change)
- An empty backup _leaves no files_. There is no `MANIFEST` or otherwise any files to such backup.
- An empty backup is expected to _leave no trace_, if possible. There will be a backup path (or the respective comparable on non-file system targets such as `s3`) created at the beginning of the process, but it is then removed once the backup is found to be empty. Such cleanup could fail, leaving an empty path with no files.
- The backup only has a _name_, used internally, during its runtime. But once found to be empty, that name is discarded.

To that effect, this PR:

- Formalizes a `BackupResult` enum (backup is unusable, is empty, or is usable)
- Distinguishes between the two scenarios: no-binlogs-need-any-backup and cannot-find-binlogs-required-for-backup. The former is now an indication of an empty backup and the process exits without error.
- Updated the PITR backup framework to test empty backup results, does not "wait-for" new backup manifest after empty backup, etc.

## Related Issue(s)

- https://github.com/vitessio/vitess/issues/11227
- Docs: https://github.com/vitessio/website/pull/1668

## Checklist

-   [ ] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [ ] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [ ] Tests were added or are not required
-   [ ] Did the new or modified tests pass consistently locally and on CI?
-   [ ] Documentation was added or is not required

### CODE REVIEW

File: go/vt/mysqlctl/backup.go [hunk 1]
```diff
@@ -168,14 +168,18 @@ func Backup(ctx context.Context, params BackupParams) error {
...
 	var finishErr error
-	if usable {
-		finishErr = bh.EndBackup(ctx)
-	} else {
+	switch backupResult {
+	case BackupUnusable:
 		logger.Errorf2(err, "backup is not usable, aborting it")
 		finishErr = bh.AbortBackup(ctx)
+	case BackupEmpty:
+		logger.Infof("backup is empty")
```
Comment (mattlord) [inline go/vt/mysqlctl/backup.go:179]: Something like `no new data to backup, skipping it` is more informative, IMO.

File: go/vt/mysqlctl/binlogs_gtid.go [hunk 2]
```diff
@@ -133,7 +132,16 @@ func ChooseBinlogsForIncrementalBackup(
 		}
 		return binaryLogsToBackup, incrementalBackupFromGTID, incrementalBackupToGTID, nil
 	}
-	return nil, "", "", vterrors.Errorf(vtrpc.Code_FAILED_PRECONDITION, "no binary logs to backup (increment is empty)")
+	if prevGTIDsUnion.Union(purgedGTIDSet).Equal(backupFromGTIDSet) {
+		// This means we've iterated over all binary logs, and as it turns out, the backup pos is
+		// identical to the Previous-GTIDs of the last binary log. But, we also know that we ourselves
+		// have flushed the binary logs so as to generate the new (now last) binary log.
+		// Which means, from the Pos of the backup till the time we issued FLUSH BINARY LOGS, there
+		// were no new GTID entries. The database was completely silent during that period.
```
Comment (mattlord) [inline go/vt/mysqlctl/binlogs_gtid.go:140]: Slightly more explanative to note in this and similar places that the database performed no writes vs "did nothing" / "was silent" etc.

File: go/vt/mysqlctl/binlogs_gtid.go [hunk 3]
```diff
@@ -133,7 +132,16 @@ func ChooseBinlogsForIncrementalBackup(
...
+		// We have nothing to backup. The backup is empty.
+		return nil, "", "", nil
+	}
+	return nil, "", "", vterrors.Errorf(vtrpc.Code_FAILED_PRECONDITION, "no binary logs to backup. backupFromGTIDSet=%v, prevGTIDsUnion=%v", backupFromGTIDSet.String(), prevGTIDsUnion.String())
```
Comment (mattlord) [inline go/vt/mysqlctl/binlogs_gtid.go:144]: It's more accurate to say that there are no new GTIDs to backup as flush logs could have been done N times (along with mysqld restarts etc). Right? Or we could say:
```
return nil, "", "", vterrors.Errorf(vtrpc.Code_FAILED_PRECONDITION, "no binary logs containing new GTIDs to backup. backupFromGTIDSet=%v, prevGTIDsUnion=%v", backupFromGTIDSet.String(), prevGTIDsUnion.String())
```

File: go/vt/mysqlctl/builtinbackupengine.go [hunk 4]
```diff
@@ -204,7 +204,7 @@ func (fe *FileEntry) open(cnf *Mycnf, readOnly bool) (*os.File, error) {
 
 // ExecuteBackup runs a backup based on given params. This could be a full or incremental backup.
 // The function returns a boolean that indicates if the backup is usable, and an overall error.
```
Comment (mattlord) [inline go/vt/mysqlctl/builtinbackupengine.go:206]: We should update the comment.

File: go/vt/mysqlctl/builtinbackupengine.go [hunk 5]
```diff
@@ -233,24 +233,25 @@ func getIncrementalFromPosGTIDSet(incrementalFromPos string) (replication.Mysql5
 // executeIncrementalBackup runs an incremental backup, based on given 'incremental_from_pos', which can be:
 // - A valid position
 // - "auto", indicating the incremental backup should begin with last successful backup end position.
-func (be *BuiltinBackupEngine) executeIncrementalBackup(ctx context.Context, params BackupParams, bh backupstorage.BackupHandle) (bool, error) {
+// The function returns a boolean that indicates if the backup is usable, and an overall error.
```
Comment (mattlord) [inline go/vt/mysqlctl/builtinbackupengine.go:236]: This comment is also off now, right?

File: go/vt/mysqlctl/builtinbackupengine.go [hunk 6]
```diff
@@ -365,14 +370,14 @@ func (be *BuiltinBackupEngine) executeIncrementalBackup(ctx context.Context, par
...
 	// In the backup manifest file, we document incrementalBackupFromGTID, not the user's requested position.
 	if err := be.backupFiles(ctx, params, bh, incrementalBackupToPosition, gtidPurged, incrementalBackupFromPosition, fromBackupName, binaryLogsToBackup, serverUUID, mysqlVersion, incrDetails); err != nil {
-		return false, err
+		return BackupUnusable, err
 	}
-	return true, nil
+	return BackupUsable, nil
 }
 
 // executeFullBackup returns a boolean that indicates if the backup is usable,
```
Comment (mattlord) [inline go/vt/mysqlctl/builtinbackupengine.go:378]: Comment off note again.

File: go/vt/mysqlctl/xtrabackupengine.go [hunk 7]
```diff
@@ -168,7 +168,7 @@ func closeFile(wc io.WriteCloser, fileName string, logger logutil.Logger, finalE
 
 // ExecuteBackup runs a backup based on given params. This could be a full or incremental backup.
 // The function returns a boolean that indicates if the backup is usable, and an overall error.
```
Comment (mattlord) [inline go/vt/mysqlctl/xtrabackupengine.go:170]: Another comment :)

File: go/vt/mysqlctl/xtrabackupengine.go [hunk 8]
```diff
@@ -177,17 +177,17 @@ func (be *XtrabackupEngine) ExecuteBackup(ctx context.Context, params BackupPara
 
 // executeFullBackup returns a boolean that indicates if the backup is usable,
```
Comment (mattlord) [inline go/vt/mysqlctl/xtrabackupengine.go:178]: You guess it

**Review by mattlord — APPROVED** [review]
LGTM! I only had minor comments and suggestions so will let you address those as you prefer.

File: go/vt/mysqlctl/binlogs_gtid.go (hunk 3 above)
Comment (shlomi-noach) [reply go/vt/mysqlctl/binlogs_gtid.go:144]: Reworded as "cannot find binary logs that cover requested GTID range"

File: go/vt/mysqlctl/binlogs_gtid.go (hunk 2 above)
Comment (shlomi-noach) [reply go/vt/mysqlctl/binlogs_gtid.go]: Reworded.

File: go/vt/mysqlctl/builtinbackupengine.go (hunk 4 above)
Comment (shlomi-noach) [reply go/vt/mysqlctl/builtinbackupengine.go]: Reworded as "The function returns a BackupResult that indicates the usability of the backup"

File: go/vt/mysqlctl/builtinbackupengine.go (hunk 5 above)
Comment (shlomi-noach) [reply go/vt/mysqlctl/builtinbackupengine.go]: Reworded

File: go/vt/mysqlctl/builtinbackupengine.go (hunk 6 above)
Comment (shlomi-noach) [reply go/vt/mysqlctl/builtinbackupengine.go]: Reworded

File: go/vt/mysqlctl/xtrabackupengine.go (hunk 7 above)
Comment (shlomi-noach) [reply go/vt/mysqlctl/xtrabackupengine.go]: Reworded

File: go/vt/mysqlctl/xtrabackupengine.go (hunk 8 above)
Comment (shlomi-noach) [reply go/vt/mysqlctl/xtrabackupengine.go]: lol

**Comment by shlomi-noach** [comment]
Docs PR: https://github.com/vitessio/website/pull/1668

**Review by vmg — APPROVED** [review]
Looks good to me!

**Pull request merged**